PPM-251 tests/run_bf.sh must return non-zero when tests fail

aciliketcap opened the pull request:
We tell bft to output test results in a JSON file and then we read that file to determine if any tests failed.

Previously bft was run with "exec", probably to have the script exit with whatever exit code bft returns. I had to remove it of course but that behavior is preserved and we return non-zero exit code when bft itself fails.

I think it is a nice addition to have boardfarm's test results in a file, may be we can use that information in other places in CI.

I used jq to parse the JSON file. I could have grepped and parsed with regular shell utils but I think jq is great to deal with JSON data and it is a must have for any sys-admin in these times.

Commit:
tests: boardfarm: make boardfarm script exit non-zero when tests fail

tests/run_bf.sh script is modified so that we can check if any tests
have failed. If there are failed tests the scripts prints the number of
failed tests in the log and exit with 1.

Note that scripts exits with 255 if something else fails but the CI job
still exits with 1.

I used jq to parse boardfarm's test report so I included it into the
boardfarm-ci docker image.

https://jira.prplfoundation.org/browse/PPM-251

Signed-off-by: Sinan Akpolat <sinan.akpolat@lifemote.com>

arnout commented on tests/run_bf.sh:
```diff
@@ -15,4 +16,9 @@ else
+        -n prplmesh_docker -x test_flows -o ${resultdir}
+
+# if bft returned nonzero return that; else return failed test count
+if [ $rc -ne 0 ]; then exit $rc; fi
```
Can't this be simplified to
```suggestion
bft -c "${bf_plugins_dir}"/boardfarm_prplmesh/prplmesh_config.json \
        -n prplmesh_docker -x test_flows -o ${resultdir} \
    || exit $?
```

You definitely are missing a `rc=$?` somewhere. As observed by shellcheck :-)

aciliketcap replied:
Yeah I need to start using a tool like shellcheck. I changed that part right before pushing it :(

aciliketcap commented:
This PR depends on https://github.com/prplfoundation/prplMesh/pull/1557 since CI tests will continue failing until that PR is merged.

RanRegev commented on tests/run_bf.sh:
```diff
@@ -15,4 +16,8 @@ else
+        -n prplmesh_docker -x test_flows -o ${resultdir} \
+	|| exit $?
+
+exit $(jq '.tests_fail' ${resultdir}/test_results.json)
```
if `tests_fail` has a non-numeric value, the entire script will have an error:
`exit: "non-numberic-value": numeric argument require`
because `exit` expect a number.
We might count on bft that it always be a number or we can test the value before directly transferring it to `exit`.

aciliketcap replied:
I changed the script to print "Unable to parse failed test count" so that it would be easier to debug if something goes wrong with test_results.json.

ghost commented:
@MergifyIO rebase

vitalii-komisarenko commented on tests/run_bf.sh:
```diff
@@ -15,4 +16,14 @@ else
+   echo "Unable to parse failed test count:" "$failed_test_count" \
+   && exit 2
+fi
+exit "$failed_test_count"
```
There are two issues with return values:
1) it is hard to tell if there was parse error or if there were exactly 2 failing tests
2) as far as I remember, only values in range 0-255 are guaranteed to be supported.

I suggest using 255 for parse error and 254 for number of failures >=254

aciliketcap replied:
Oh, that makes sense. I added 2 because boardfarm returns 2 when it has a problem.

I am updating the PR as you suggested.

aciliketcap replied:
I also changed the exit code when boardfarm fails to 255 too.

Let's try to use it this way for a while, we can make changes if we think that we need actual exit codes instead of failed test count.

vitalii-komisarenko commented on tests/run_bf.sh:
```diff
@@ -15,4 +16,14 @@ else
+   echo "Unable to parse failed test count:" "$failed_test_count" \
+   && exit 255
+fi
+exit "$failed_test_count"
```
On my ubuntu machine `exit 256` actually returns 0.
Not sure if we are going to fail 256 tests at once, but technically there is integer overflow.

aciliketcap replied:
I don't think we will have that many tests but I'll do the necessary check in bash.

aciliketcap replied:
I fixed max failed test count to return to 254

rmelotte committed:
.gitlab-ci.yml: boardfarm: always upload the artifacts

With Gitlab CI by default the artifacts are only uploaded if the job
succeeds.

Add the "when" parameters to the boardfarm artifacts, to always upload
them regardless of the job's status.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

aciliketcap commented:
I wanted the CI job to return the failed test count but as far as I investigated GitLab CI job exits with 1 when the script in the job exits with non-zero value. So we exit with 1 when there are failed tests and print the number of failed tests at the bottom of the log.
Another issue was "[Errno 2] No such file or directory: './results/test_results.json'" message in logs but it is printed by boardfarm and is not really related to our scripts or tests so I am not doing anything about it.

rmelotte reviewed (approved):
Thanks :+1: 
To me we can disable the 2 failing tests here and open another PR later to fix them.
This way in the meantime, the boardfarm job would finally do something useful (checking that the 2 passing ones keep on passing).

aciliketcap committed:
tests: boardfarm: removed two tests which were failing in GitLab CI

ApConfigRenew and LinkMetricQuery tests were failing. They will be fixed
and reenabled in PPM-331.

Signed-off-by: Sinan Akpolat <aciliketcap@gmail.com>
